tests/beta/interfaces/test_tape_torch.py

```diff
@@ -0,0 +1,447 @@
+# Copyright 2018-2020 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Unit tests for the torch interface"""
+import pytest
+
+torch = pytest.importorskip("torch", minversion="1.3")
+
+import numpy as np
+
+import pennylane as qml
+from pennylane.beta.tapes import QuantumTape
+from pennylane.beta.queuing import expval, var, sample, probs
+from pennylane.beta.interfaces.torch import TorchInterface
+
+
+class TestTorchQuantumTape:
+    """Test the autograd interface applied to a tape"""
+
+    def test_interface_construction(self):
+        """Test that the interface is correctly applied"""
+        with TorchInterface.apply(QuantumTape()) as tape:
+            qml.RX(0.5, wires=0)
+            expval(qml.PauliX(0))
+
+        assert tape.interface == "torch"
+        assert isinstance(tape, TorchInterface)
+        assert tape.__bare__ == QuantumTape
+
+    def test_repeated_interface_construction(self):
+        """Test that the interface is correctly applied multiple times"""
+        with TorchInterface.apply(QuantumTape()) as tape:
+            qml.RX(0.5, wires=0)
+            expval(qml.PauliX(0))
+
+        assert tape.interface == "torch"
+        assert isinstance(tape, TorchInterface)
+        assert tape.__bare__ == QuantumTape
+
+        TorchInterface.apply(QuantumTape())
+        assert tape.interface == "torch"
+        assert isinstance(tape, TorchInterface)
+        assert tape.__bare__ == QuantumTape

```

Comment:
Not sure that `tape` has been reassigned at all

Reply:
This is because `TorchInterface.apply` acts in-place, modifying the tape argument it is given.

I originally tried to implement a non-in-place method, of the form:

```python
new_tape = TorchInterface.apply(original_tape)
```
(which I think is more functional).

_But_, this requires doing a deep copy of the original tape, which works fine in autograd and TF, but not pytorch (torch tensors that require gradients do not support copying). So I changed this method to be in-place across the board.

Which, in hindsight, might be better/more efficient, developers just have to be aware of the side effect.

Reply:
what about `XYZInterface.replace(new_tape)` to be more explicit?

Reply:
or maybe it's just that you didn't bother writing any ops on the new tape that confused me

Reply:
Oops, no you're right, this is my mistake. The test has a typo, it should be
```python
        assert tape.interface == "torch"
        assert isinstance(tape, TorchInterface)
        assert tape.__bare__ == QuantumTape

        TorchInterface.apply(tape)
        assert tape.interface == "torch"
        assert isinstance(tape, TorchInterface)
        assert tape.__bare__ == QuantumTape
```

I'll fix this, and add an extra assertion just to make sure it _is_ the same tape in both cases.